src/nacl/public.py

```diff
@@ -92,17 +92,18 @@ def __init__(self, private_key, encoder=encoding.RawEncoder):
 
         self._private_key = private_key
         self.public_key = PublicKey(raw_public_key)
+        self._hashbytes = b'p:' + bytes(self.public_key)
 
     def __bytes__(self):
         return self._private_key
 
     def __hash__(self):
-        return hash(bytes(self))
+        return hash(self._hashbytes)
```

Comment:
The change in itself is to let equivalent PrivateKeys check as equal, while keeping the property (I feel desirable, even if a little overkill) that a PrivateKey's hash and its corresponding PublicKey's are different. If you think so, I could just as well directly return hash(bytes(self.public_key)) or hash((self.__class__, bytes(self.public_key))).